File: src/FluentCards/ColumnBuilder.cs
namespace FluentCards;

/// <summary>
/// Fluent builder for creating Column elements.
/// </summary>
public class ColumnBuilder
{
    private readonly Column _column = new() { Items = new List<AdaptiveElement>() };

    /// <summary>
    /// Sets the width of the column.
    /// </summary>
    /// <param name="width">The column width (e.g., "auto", "stretch", "50px", or "2").</param>
    /// <returns>The builder instance for method chaining.</returns>
    public ColumnBuilder WithWidth(string width)
    {
        _column.Width = width;
        return this;
    }

    /// <summary>
    /// Sets the unique identifier for the column.
    /// </summary>
    /// <param name="id">The unique identifier.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public ColumnBuilder WithId(string id)
    {
        _column.Id = id;
        return this;
    }

    /// <summary>
    /// Sets the style of the column.
    /// </summary>
    /// <param name="style">The container style.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public ColumnBuilder WithStyle(ContainerStyle style)
    {
        _column.Style = style;
        return this;
    }

    /// <summary>
    /// Sets the vertical alignment of content within the column.
    /// </summary>
    /// <param name="alignment">The vertical alignment.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public ColumnBuilder WithVerticalContentAlignment(VerticalAlignment alignment)
    {
        _column.VerticalContentAlignment = alignment;
        return this;
    }

    /// <summary>
    /// Enables or disables bleed for the column.
    /// </summary>
    /// <param name="bleed">True to enable bleed, false to disable.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public ColumnBuilder WithBleed(bool bleed = true)
    {
        _column.Bleed = bleed;
        return this;
    }

    /// <summary>
    /// Sets the minimum height of the column.
    /// </summary>
    /// <param name="minHeight">The minimum height.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public ColumnBuilder WithMinHeight(string minHeight)
    {
        _column.MinHeight = minHeight;
        return this;
    }

    /// <summary>
    /// Configures the background image for the column.
    /// </summary>
    /// <param name="configure">Action to configure the background image.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public ColumnBuilder WithBackgroundImage(Action<BackgroundImageBuilder> configure)
    {
        var builder = new BackgroundImageBuilder();
        configure(builder);
        _column.BackgroundImage = builder.Build();
        return this;
    }

    /// <summary>
    /// Configures the action to invoke when the column is selected.
    /// </summary>
    /// <param name="configure">Action to configure the select action.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public ColumnBuilder WithSelectAction(Action<ActionBuilder> configure)
    {
        var builder = new ActionBuilder();
        configure(builder);
        _column.SelectAction = builder.Build();
        return this;
    }

    /// <summary>
    /// Adds a TextBlock to the column.
    /// </summary>
    /// <param name="configure">Action to configure the TextBlock.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public ColumnBuilder AddTextBlock(Action<TextBlockBuilder> configure)
    {
        var builder = new TextBlockBuilder();
        configure(builder);
        _column.Items!.Add(builder.Build());
        return this;
    }

    /// <summary>
    /// Adds an Image to the column.
    /// </summary>
    /// <param name="configure">Action to configure the Image.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public ColumnBuilder AddImage(Action<ImageBuilder> configure)
    {
        var builder = new ImageBuilder();
        configure(builder);
        _column.Items!.Add(builder.Build());
        return this;
    }

    /// <summary>
    /// Adds a Container to the column.
    /// </summary>
    /// <param name="configure">Action to configure the Container.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public ColumnBuilder AddContainer(Action<ContainerBuilder> configure)
    {
        var builder = new ContainerBuilder();
        configure(builder);
        _column.Items!.Add(builder.Build());
        return this;
    }

    /// <summary>
    /// Sets the fallback behavior for the column.
    /// </summary>
    /// <param name="fallback">The fallback value ("drop" or another element).</param>
    /// <returns>The builder instance for method chaining.</returns>
    public ColumnBuilder WithFallback(object fallback)
    {
        _column.Fallback = fallback;
        return this;
    }

    /// <summary>
    /// Sets the height of the column.
    /// </summary>
    /// <param name="height">The height ("auto" or "stretch").</param>
    /// <returns>The builder instance for method chaining.</returns>
    public ColumnBuilder WithHeight(string height)
    {
        _column.Height = height;
        return this;
    }

    /// <summary>
    /// Sets whether a separator line is drawn at the left of the column.
    /// </summary>
    /// <param name="separator">True to show separator.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public ColumnBuilder WithSeparator(bool separator = true)
    {
        _column.Separator = separator;
        return this;
    }

    /// <summary>
    /// Sets the spacing between this column and the preceding column.
    /// </summary>
    /// <param name="spacing">The spacing value.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public ColumnBuilder WithSpacing(Spacing spacing)
    {
        _column.Spacing = spacing;
        return this;
    }

    /// <summary>
    /// Sets whether the column is visible.
    /// </summary>
    /// <param name="isVisible">True if visible, false if hidden.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public ColumnBuilder WithIsVisible(bool isVisible)
    {
        _column.IsVisible = isVisible;
        return this;
    }

    /// <summary>
    /// Sets the feature requirements for the column.
    /// </summary>
    /// <param name="key">The feature key.</param>
    /// <param name="version">The minimum version required.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public ColumnBuilder WithRequires(string key, string version)
    {
        _column.Requires ??= new Dictionary<string, string>();
        _column.Requires[key] = version;
        return this;
    }

    /// <summary>
    /// Sets whether content in this column should be presented right to left.
    /// </summary>
    /// <param name="rtl">True for right-to-left.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public ColumnBuilder WithRtl(bool rtl = true)
    {
        _column.Rtl = rtl;
        return this;
    }

    /// <summary>
    /// Builds and returns the configured Column.
    /// </summary>
    /// <returns>The configured Column instance.</returns>
    public Column Build()
    {
        return _column;
    }
}

File: go/fluentcards/column_builder.go
package fluentcards

// ColumnBuilder builds a Column element within a ColumnSet.
type ColumnBuilder struct {
	data map[string]any
}

func newColumnBuilder() *ColumnBuilder {
	return &ColumnBuilder{data: map[string]any{"type": "Column", "items": []any{}}}
}

func (b *ColumnBuilder) WithID(id string) *ColumnBuilder {
	b.data["id"] = id
	return b
}

func (b *ColumnBuilder) WithWidth(width string) *ColumnBuilder {
	b.data["width"] = width
	return b
}

func (b *ColumnBuilder) WithStyle(style ContainerStyle) *ColumnBuilder {
	b.data["style"] = string(style)
	return b
}

func (b *ColumnBuilder) WithVerticalContentAlignment(alignment VerticalAlignment) *ColumnBuilder {
	b.data["verticalContentAlignment"] = string(alignment)
	return b
}

func (b *ColumnBuilder) WithBleed(bleed bool) *ColumnBuilder {
	b.data["bleed"] = bleed
	return b
}

func (b *ColumnBuilder) WithMinHeight(minHeight string) *ColumnBuilder {
	b.data["minHeight"] = minHeight
	return b
}

func (b *ColumnBuilder) WithBackgroundImage(configure func(*BackgroundImageBuilder)) *ColumnBuilder {
	bib := newBackgroundImageBuilder()
	configure(bib)
	b.data["backgroundImage"] = bib.Build()
	return b
}

func (b *ColumnBuilder) WithSelectAction(configure func(*ActionBuilder)) *ColumnBuilder {
	ab := newActionBuilder()
	configure(ab)
	b.data["selectAction"] = ab.Build()
	return b
}

// WithIsVisible sets whether the column is visible.
func (b *ColumnBuilder) WithIsVisible(isVisible bool) *ColumnBuilder {
	b.data["isVisible"] = isVisible
	return b
}

// WithSpacing sets the spacing between this column and the preceding column.
func (b *ColumnBuilder) WithSpacing(spacing Spacing) *ColumnBuilder {
	b.data["spacing"] = string(spacing)
	return b
}

// WithSeparator sets whether a separator line is drawn at the left of the column.
func (b *ColumnBuilder) WithSeparator(separator bool) *ColumnBuilder {
	b.data["separator"] = separator
	return b
}

// WithHeight sets the height of the column ("auto" or "stretch").
func (b *ColumnBuilder) WithHeight(height string) *ColumnBuilder {
	b.data["height"] = height
	return b
}

// WithFallback sets the fallback behavior when the column is unsupported.
func (b *ColumnBuilder) WithFallback(fallback any) *ColumnBuilder {
	b.data["fallback"] = fallback
	return b
}

// WithRequires sets a feature requirement for the column.
func (b *ColumnBuilder) WithRequires(key, version string) *ColumnBuilder {
	reqs, ok := b.data["requires"].(map[string]any)
	if !ok {
		reqs = map[string]any{}
	}
	reqs[key] = version
	b.data["requires"] = reqs
	return b
}

// WithRtl sets whether content should be presented right to left.
func (b *ColumnBuilder) WithRtl(rtl bool) *ColumnBuilder {
	b.data["rtl"] = rtl
	return b
}

func (b *ColumnBuilder) AddTextBlock(configure func(*TextBlockBuilder)) *ColumnBuilder {
	tb := newTextBlockBuilder()
	configure(tb)
	b.data["items"] = append(b.data["items"].([]any), tb.Build())
	return b
}

func (b *ColumnBuilder) AddImage(configure func(*ImageBuilder)) *ColumnBuilder {
	ib := newImageBuilder()
	configure(ib)
	b.data["items"] = append(b.data["items"].([]any), ib.Build())
	return b
}

func (b *ColumnBuilder) AddContainer(configure func(*ContainerBuilder)) *ColumnBuilder {
	cb := newContainerBuilder()
	configure(cb)
	b.data["items"] = append(b.data["items"].([]any), cb.Build())
	return b
}

func (b *ColumnBuilder) AddElement(element Card) *ColumnBuilder {
	b.data["items"] = append(b.data["items"].([]any), element)
	return b
}

func (b *ColumnBuilder) Build() Card {
	return b.data
}
